src/protocols/bdx/BdxMessages.cpp

```diff
@@ -0,0 +1,456 @@
+    {
+        aBuffer.Put(FileDesignator, static_cast<size_t>(FileDesLength));
+    }
+
+    if (Metadata)
+    {
+        aBuffer.Put(Metadata, static_cast<size_t>(MetadataLength));
+    }
+}
+
+CHIP_ERROR TransferInit::Parse(const System::PacketBuffer & aBuffer)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    uint8_t proposedTransferCtl;
+    uint8_t rangeCtl;
+    uint32_t tmpUint32Value = 0; // Used for reading non-wide length and offset fields
+    uint8_t * bufStart      = aBuffer.Start();
+    Reader bufReader(bufStart, aBuffer.DataLength());
+    BitFlags<uint8_t, RangeControlFlags> rangeCtlFlags;
+
+    SuccessOrExit(bufReader.Read8(&proposedTransferCtl).Read8(&rangeCtl).Read16(&MaxBlockSize).StatusCode());
+
+    Version = proposedTransferCtl & kVersionMask;
+    TransferCtlOptions.SetRaw(static_cast<uint8_t>(proposedTransferCtl & ~kVersionMask));
+    rangeCtlFlags.SetRaw(rangeCtl);
+
+    if (rangeCtlFlags.Has(kStartOffset))
+    {
+        if (rangeCtlFlags.Has(kWiderange))
+        {
+            SuccessOrExit(bufReader.Read64(&StartOffset).StatusCode());
+        }
+        else
+        {
+            SuccessOrExit(bufReader.Read32(&tmpUint32Value).StatusCode());
+            StartOffset = tmpUint32Value;
+        }
+    }
+
+    if (rangeCtlFlags.Has(kDefLen))
+    {
+        if (rangeCtlFlags.Has(kWiderange))
+        {
+            SuccessOrExit(bufReader.Read64(&MaxLength).StatusCode());
+        }
+        else
+        {
+            SuccessOrExit(bufReader.Read32(&tmpUint32Value).StatusCode());
+            MaxLength = tmpUint32Value;
+        }
+    }
+
+    SuccessOrExit(bufReader.Read16(&FileDesLength).StatusCode());
+
+    // WARNING: this struct will store a pointer to the start of the file designator in the PacketBuffer,
+    // but will not make a copy. It is essential that this struct not outlive the PacketBuffer,
+    // or there is risk of unsafe memory access.
+    VerifyOrExit(bufReader.HasAtLeast(FileDesLength), err = CHIP_ERROR_MESSAGE_INCOMPLETE);
+    FileDesignator = &bufStart[bufReader.OctetsRead()];
+
+    // Rest of message is metadata (could be empty)
+    if (bufReader.Remaining() > FileDesLength)
+    {
+        // WARNING: this struct will store a pointer to the start of metadata in the PacketBuffer,
+        // but will not make a copy. It is essential that this struct not outlive the PacketBuffer,
+        // or there is risk of unsafe memory access.
+        uint16_t metadataStartIndex = static_cast<uint16_t>(bufReader.OctetsRead() + FileDesLength);
+        Metadata                    = &bufStart[metadataStartIndex];
+        MetadataLength              = static_cast<uint16_t>(aBuffer.DataLength() - metadataStartIndex);
+    }
+
+exit:
+    if (bufReader.StatusCode() != CHIP_NO_ERROR)
+    {
+        err = bufReader.StatusCode();
+    }
+    return err;
+}
+
+size_t TransferInit::PackedSize() const
+{
+    System::PacketBuffer * pbuf = System::PacketBuffer::NewWithAvailableSize(0);
```

Comment:
You do not need a real packet buffer here. You could use:

```
  uint8_t dummy[0];
  BufBound emptyBuf(&dummy[0], sizeof(dummy));
```

This avoids passing a `nullptr` to `BufBound`, while at same time avoiding creation of a new packet buffer.

Same comment for all other `PackedSize()`

Reply:
+1 NewWithAvailable may actually over-allocate. Static storage is better.

Also Pack is in theory allowed to early return. You need to add a comment that it is specifically not allowed to early return.

Reply:
Agreed this is better than using a PacketBuffer

Reply:
@andy31415 I'll add comments to the .cpp file